hw2

cyberpanncake opened the pull request.

cyberpanncake committed:
feat: add clients and scheduler

cyberpanncake committed:
feat: add tests and some fixes to github client

cyberpanncake committed:
feat: add tests and some fixes to stackoverflow client

strelchm commented on scrapper/src/main/java/edu/java/client/AbstractClient.java:
```diff
@@ -0,0 +1,14 @@
+    protected WebClient client;
+
+    public AbstractClient(String baseUrl) {
+        if (baseUrl.isEmpty()) {
```
Валидация baseUrl должна быть на уровне пропертей - если пропертя от клиента пустая, то используем дефолтный URI, который захардкоженый. см комментарий в ClientConfig

strelchm commented on scrapper/src/main/java/edu/java/configuration/ClientConfig.java:
```diff
@@ -0,0 +1,20 @@
+public class ClientConfig {
+
+    @Bean
+    public GithubClient githubClient() {
```
По заданию:

> При создании клиента обязательно должна быть возможность указать базовый URL. При этом если он не указывается, то должен использоваться URL по умолчанию.

В текущей реализации берется из пропертей всегда URL. Предлагаю переделать следующим образом:

1.  Создать класс конфиг пропертей по подобию edu.java.bot.configuration.ApplicationConfig с базовыми URI (либо не создать, а добавить в текущую конфигурацию ApplicationConfig)
2. Инжектить конфиг пропертей в данный конфиг
3. Если URI из пропертей пустой, то использовать дефолтный (захардкожен в коде)

cyberpanncake replied:
Заинжектила проперти с базовыми URL в ClientConfig. При создании клиента, если URL пустой, то подставляется захардкоженная строка. Скажите, пожалуйста, такая реализация подойдёт? В AbstractClient проверку на пустой URL с бросанием исключения оставила, т.к. на уровне этого класса непонятно, каким должен быть дефолтный URL

strelchm replied:
да, сейчас хорошо, те клиент если в конфигах не будет значения - берем дефолтное

strelchm replied:
только похоже что сейчас мб NPE на месте isEmpty, лучше так:
```
    @Bean
    public GithubClient githubClient() {
        return new GithubClient(githubUrl, Strings.isNotEmpty(githubUrl) ? githubUrl : DEFAULT_GH_URL);
    }
```

strelchm replied:
Касательно AbstractClient - предполагается, что в него не будет ни в каком случае передано null-значения. Те, возможно, избыточно прокидывание исключения. Для таких целей подстраховки подойдет аннотация ломбока - которая сгенирирует перед коспайлом проверку на null с прокидыванием исключения:
```
package edu.java.client;

import lombok.NonNull;
import org.springframework.web.reactive.function.client.WebClient;

public abstract class AbstractClient {
    protected WebClient client;

    public AbstractClient(@NonNull String baseUrl) {
        this.client = WebClient.builder().baseUrl(baseUrl).build();
    }
}
```

strelchm commented on scrapper/src/main/java/edu/java/dto/GithubResponse.java:
```diff
@@ -0,0 +1,24 @@
+    Actor actor,
+    Repository repo,
+    @JsonProperty("created_at")
+    OffsetDateTime createdAt
```
nit: По-дефолту сериализация/десериализация в Camel case, можно установить сериализацию/десериализацию в пропертях и убрать @JsonProperty:
```
spring.jackson.property-naming-strategy=SNAKE_CASE
```
Либо если отличается какое-то DTO, то над классом правило устанавливать:
```
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
```

strelchm commented on scrapper/src/test/java/edu/java/client/GithubClientTest.java:
```diff
@@ -0,0 +1,119 @@
+            List<GithubResponse> responses = objectMapper.readValue(json, new TypeReference<>(){});
+            return responses.get(0);
+        } catch (Exception e) {
+            return null;
```
Странно, что в контракте, по которому возвращаем Optional - возвращаем **null**. Возможно, **Optional.empty()** будет лучше?

cyberpanncake replied:
Я здесь использовала Optional, чтобы отделить "хороший" null (когда мы запросили обновление, а его нет), от "плохого" (когда при запросе что-то пошло не так). Тогда наверное лучше будет возврат null заменить на бросание собственного исключения

strelchm replied:
кажется, что тогда исключение будет более явным - возникла ошибка

strelchm commented on scrapper/src/main/java/edu/java/client/StackoverflowClient.java:
```diff
@@ -0,0 +1,49 @@
+            .retrieve()
+            .bodyToMono(JsonNode.class);
+        try {
+            JsonNode root = jsonNodeMono.block();
```
Править не обязательно: обычно можно решить проблемы десериализации с помощью Jackson, который под коробкой в Spring / WebClient. Можно, например, [кастомный десериализатор](https://www.baeldung.com/jackson-deserialization#on-mapper) написать. Кажется, что здесь можно сразу тип в bodyToMono не JsonNode сделать, а StackoverflowResponse

strelchm commented on scrapper/src/main/java/edu/java/client/GithubClient.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public Optional<GithubResponse> getUpdate(String author, String repository) {
+        String request = String.format("networks/%s/%s/events", author, repository);
```
может наполнение path-переменными также [перенести ](https://www.baeldung.com/webflux-webclient-parameters#uri-path-components)в path() ?

strelchm commented on scrapper/src/main/java/edu/java/client/GithubClient.java:
```diff
@@ -0,0 +1,38 @@
+                .retrieve()
+                .bodyToMono(GithubResponse[].class)
+                .block();
+            if (responses == null || responses.length == 0) {
```
м б так:
```
return Stream.ofNullable(responses)
    .flatMap(Arrays::stream)
    .findFirst();
```

strelchm commented on scrapper/src/main/java/edu/java/client/StackoverflowClient.java:
```diff
@@ -0,0 +1,49 @@
+    private static final String DEFAULT_URL = new Properties().getProperty("app.base-url.stackoverflow");
+    private final static ObjectMapper MAPPER = new ObjectMapper();
+
+    static {
```
Почему здесь используем не-дефолтный мапер? получается, что мы  `client.get().uri(uriBuilder -> uriBuilder ...` используем один маппер, который дефолтный, а после пользуемся другим. Может сконфигурируем дефолтный маппер и сюда его заинжектим?

cyberpanncake replied:
Я наверное не совсем Вас поняла, но в uriBuilder маппера как будто бы и нет в явном виде (внутри он может где-то и есть). А ObjectMapper MAPPER я использую для преобразования JsonNode в StackoverflowResponse. И кроме этого места одного нигде его не использую. Поэтому я не очень понимаю, что мне нужно сделать

strelchm replied:
Через new в spring так себе объекты создавать, лучше инжектить бины - инверсия зависимости. Здесь лучше инжектить дефолтный маппер:
```
@Configuration
@RequiredArgsConstructor
public class ClientConfig {
    @Value("${app.base-url.github}")
    private final String githubUrl;
    @Value("${app.base-url.stackoverflow}")
    private final String stackoverflowUrl;
    private final ObjectMapper objectMapper;

    @Bean
    public GithubClient githubClient() {
        return new GithubClient(githubUrl, objectMapper);
    }

    @Bean
    public StackoverflowClient stackoverflowClient(ObjectMapper objectMapper) {
        return new StackoverflowClient(stackoverflowUrl, objectMapper);
    }
}
```
и его донастроить, т к модуль времени в нем не помешает:
```
package edu.java.configuration;

import com.fasterxml.jackson.databind.Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class JacksonModuleConfig {
    @Bean
    public Module javaTimeModule() {
        return new JavaTimeModule();
    }
}
```
Сверху [зарегистрировали ](https://www.baeldung.com/spring-boot-customize-jackson-objectmapper#customizing-the-default-objectmapper)модуль для дефолтного маппера. В итоге получили один маппер, который в разных местах будет настраивать сер/десер времени, а не много мапперов по разному настроеных в разных местах

strelchm commented on scrapper/src/main/java/edu/java/client/GithubClient.java:
```diff
@@ -0,0 +1,38 @@
+        super(baseUrl);
+    }
+
+    public Optional<GithubResponse> getUpdate(String author, String repository) {
```
Предлагаю вместо Optional возвращать Mono (в mono лежит либо 0 либо 1), а уже block() делать извне клиента, останется только bodyToMono
Mono - это что-то вроде Optional в реактивном программировании. Код может использоваться из реактивного стека

cyberpanncake replied:
К сожалению, у меня не получилось переделать Optional на Mono. Всё осложняется тем, что в ответе от github лежит массив и нужно как-то до block() вытащить его первый элемент, а в ответе от stackoverflow вложенная структура: в поле items лежит массив, из которого тоже надо получить первый элемент до block(). На данный момент мне не хватает знаний, чтобы понять, как это реализовать.
Я бы хотела Вас попросить подсказать мне немного, как это может примерно выглядеть, тогда я ещё раз попробую переделать на Mono

strelchm replied:
Как насчет такого:
```
    public Mono<GithubResponse> getUpdate(String author, String repository) throws ResponseException {
        try {
            return client.get()
                .uri(uriBuilder -> uriBuilder
                    .path("/repos/{author}/{repo}/events")
                    .queryParam("per_page", 1)
                    .build(author, repository))
                .retrieve()
                .bodyToMono(GithubResponse[].class)
                .filter(githubResponses -> githubResponses.length > 0)
                .mapNotNull(githubResponses -> githubResponses[0]);
        } catch (WebClientResponseException e) {
            throw new ResponseException();
        }
    }
```

strelchm replied:
т е bodyToMono возвращает объект Mono, с ним +- мы можем делать преобразования как с Optional. block будем извне вызывать

strelchm replied:
в таком подходе как описал выше - перехват исключения будет валидным только во время вызова block. Поэтому, что бы корректно конвертировать ошибку, подход с mono будет выглядить так:
```
    public Mono<GithubResponse> getUpdate(String author, String repository) throws ResponseException {
        return client.get()
            .uri(uriBuilder -> uriBuilder
                .path("/repos/{author}/{repo}/events")
                .queryParam("per_page", 1)
                .build(author, repository))
            .retrieve()
            .bodyToMono(GithubResponse[].class)
            .filter(githubResponses -> githubResponses.length > 0)
            .mapNotNull(githubResponses -> githubResponses[0])
            .onErrorResume(e -> Mono.error(e instanceof WebClientResponseException ? new ResponseException() : e));
    }
```
Во время вызова block при WebClientResponseException будет кидаться ResponseException

strelchm replied:
эти правки опциональны, т к касаются реактивного программирования, но на будущее будет полезно. Можно оставить так, как сейчас есть

cyberpanncake replied:
Спасибо, стало понятнее)

cyberpanncake committed:
fix: base url validation in ClientConfig

cyberpanncake committed:
fix: dto fields naming without @JsonProperty

cyberpanncake committed:
fix:
1. throw ResponseException instead of null in clients
2. move path vars to path in uri builders
3. get 1 of github responses with stream

cyberpanncake committed:
fix: make custom object mapper as bean

strelchm commented on scrapper/src/main/java/edu/java/dto/StackoverflowResponse.java:
```diff
@@ -0,0 +1,22 @@
+import java.time.OffsetDateTime;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public record StackoverflowResponse(
```
в случае camel case:
```
package edu.java.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import java.time.OffsetDateTime;

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record StackoverflowResponse(
    Owner owner,
    OffsetDateTime last_activityDate,
    Long answerId,
    Long questionId
) {
    @JsonIgnoreProperties(ignoreUnknown = true)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Owner(
        String displayName
    ) {
    }
}
```

cyberpanncake replied:
Вы говорили, что можно либо аннотацию @JsonNaming указать, либо в properties прописать стратегию. Я это как раз в properties указала:
spring:
  application:
    name: scrapper
  jackson:
    property-naming-strategy: SNAKE_CASE

strelchm replied:
понял, были переменные в таком виде display_name, они в переменные camel case через оба способа должны маппиться, без надобности переменную именовать в camel case

cyberpanncake replied:
Всё, до меня теперь дошло. Была невнимательна, когда прошлый вариант писала, и не подумала о том, что если прописать стратегию нейминга, то имена полей можно оставить в camelCase и они автоматически в snake_case смаппятся, потому что для этого и нужна стратегия) Переделала поля на camelCase

strelchm replied:
Теперь круто, спасибо

strelchm replied:
![image](https://github.com/cyberpanncake/Torzhkova-Tinkoff-JavaBackendCourse-Spring/assets/23243577/43be3f1f-d4f0-44ef-8edb-9071f91894ee)
Похоже, что судя по тестам не работает пропертя из настроек из-за переопределения маппера - значения все навыходе в dto null-евые, лечится таким явным заданием

cyberpanncake replied:
Спасибо! Сейчас поправлю

strelchm commented on scrapper/src/main/java/edu/java/configuration/ClientConfig.java:
```diff
@@ -2,30 +2,36 @@
     private String githubUrl;
     @Value("${app.base-url.stackoverflow}")
     private String stackoverflowUrl;
+    private final ObjectMapperConfig mapperConfig;
```
Здесь лучше маппить ObjectMapper objectMapper, как в комменте - класс конфига это по сути обертка для создания бинов из методов с помощью аннотации @Bean

cyberpanncake committed:
fix: inject mapper instead of mapper config in ClientConfig

cyberpanncake committed:
fix: rename dto fields

cyberpanncake committed:
fix: json to response mapping
